### PR TITLE
Fix #25131: Template Builder integrate Box Component in Row Component

### LINKED ISSUE #25131: Template Builder: Integrate Box Components to Row Component
### User Story
As a dotCMS stakeholder, I want to be able to add the box components to the template builder row component.

### Acceptance Criteria
- Integration of box components to the template builder row component
- The box component should be resizable
- The box component should be able to move to different rows and columns
- The box component should change based on the columns width

### Proposed Objective
Core Features

### Proposed Priority
Priority 3 - Average

### External Links... Slack Conversations, Support Tickets, Figma Designs, etc.:
N/A

### Assumptions & Initiation Needs
- Prior understanding of template builder row component and box components

### Quality Assurance Notes & Workarounds
N/A

### Sub-Tasks & Estimates:
- Integrate box components to the template builder row component)
- Add and remove box components from the template builder row component
- Resize box components within the row component

### PR DESCRIPTION
**Pull request by zJaaal**
### Summary
🎨🛠️🧪

This pull request enhances the template builder component and its related components, services, and utilities. It adds a new feature to remove columns from the grid, and refactors the box size logic to use a numeric value instead of an enum. It also improves the layout, readability, and testing of the template builder code. It removes some unused or redundant styles and files, and adds some attributes and dependencies for testing and storybook.

> _`TemplateBuilder` grows_
> _Adding, removing columns_
> _Autumn of refactor_

### Walkthrough
*  Rename `size` input property of `TemplateBuilderBoxComponent` to `boxSize` and change its type from `TemplateBuilderBoxSize` to `number` ([link](https://github.com/dotCMS/core/pull/25138/files?diff=unified&w=0#diff-bad3f14381a1639d24dc3958319e2ffaf13ad3342ea3276184e86b7fbd4c3e09L3-R6), [link](https://github.com/dotCMS/core/pull/25138/files?diff=unified&w=0#diff-bad3f14381a1639d24dc3958319e2ffaf13ad3342ea3276184e86b7fbd4c3e09L12-R12), [link](https://github.com/dotCMS/core/pull/25138/files?diff=unified&w=0#diff-769c98d8aaa736f86942f3d9d20765bb660348e4938b8911baf548f0572f94c7L8-R8), [link](https://github.com/dotCMS/core/pull/25138/files?diff=unified&w=0#diff-769c98d8aaa736f86942f3d9d20765bb660348e4938b8911baf548f0572f94c7L28-R25), [link](https://github.com/dotCMS/core/pull/25138/files?diff=unified&w=0#diff-769c98d8aaa736f86942f3d9d20765bb660348e4938b8911baf548f0572f94c7L46-R43), [link](https://github.com/dotCMS/core/pull/25138/files?diff=unified&w=0#diff-769c98d8aaa736f86942f3d9d20765bb660348e4938b8911baf548f0572f94c7L54-R51), [link](https://github.com/dotCMS/core/pull/25138/files?diff=unified&w=0#diff-769c98d8aaa736f86942f3d9d20765bb660348e4938b8911baf548f0572f94c7L62-R59), [link](https://github.com/dotCMS/core/pull/25138/files?diff=unified&w=0#diff-769c98d8aaa736f86942f3d9d20765bb660348e4938b8911baf548f0572f94c7L71-R68), [link](https://github.com/dotCMS/core/pull/25138/files?diff=unified&w=0#diff-522ebd82bd93c99880b22d8a3dddaf00afd05b8814a1d6c969f121493333014aL35-R45), [link](https://github.com/dotCMS/core/pull/25138/files?diff=unified&w=0#diff-1fde9398eccbbd88fe2b6ce2c5d40bcefeb93c5bc3c7f110e3a1ed6b29b32aa3L40-R50))
*  Implement `OnChanges` interface in `TemplateBuilderBoxComponent` and add `boxSize` property to store the computed value of the box size based on the `boxSize` input ([link](https://github.com/dotCMS/core/pull/25138/files?diff=unified&w=0#diff-522ebd82bd93c99880b22d8a3dddaf00afd05b8814a1d6c969f121493333014aL2-R16), [link](https://github.com/dotCMS/core/pull/25138/files?diff=unified&w=0#diff-522ebd82bd93c99880b22d8a3dddaf00afd05b8814a1d6c969f121493333014aL23-R27), [link](https://github.com/dotCMS/core/pull/25138/files?diff=unified&w=0#diff-522ebd82bd93c99880b22d8a3dddaf00afd05b8814a1d6c969f121493333014aL35-R45))
*  Add `TemplateBuilderBoxComponent` to the template of `TemplateBuilderComponent` and pass the `items`, `boxSize`, and `deleteColumn` properties as inputs ([link](https://github.com/dotCMS/core/pull/25138/files?diff=unified&w=0#diff-1fde9398eccbbd88fe2b6ce2c5d40bcefeb93c5bc3c7f110e3a1ed6b29b32aa3L40-R50))
*  Add `TemplateBuilderBoxSize` enum to the `models.ts` file and use it in the `getBoxSize` function in the `gridstack-utils.ts` file to return the size of a box based on the number of width ([link](https://github.com/dotCMS/core/pull/25138/files?diff=unified&w=0#diff-b93c6205d24c66efa12727a86a6616d0bd0fdcd030871cd4ead5d72a69792235R31-R36), [link](https://github.com/dotCMS/core/pull/25138/files?diff=unified&w=0#diff-f3792a46b295fb618b6d4a8d8f4ecef72ec5e3c41a470d6a44886881a6b93659L5-R5), [link](https://github.com/dotCMS/core/pull/25138/files?diff=unified&w=0#diff-f3792a46b295fb618b6d4a8d8f4ecef72ec5e3c41a470d6a44886881a6b93659R136-R149))
*  Rename `updateColumn` method of `DotTemplateBuilderStore` service to `updateColumnGridStackData` and simplify its logic to only update the `x`, `y`, and `w` properties of the column ([link](https://github.com/dotCMS/core/pull/25138/files?diff=unified&w=0#diff-31b7a396b598816d0a94599edd0c1338d80f276fed9775642769710040776e24L180-R180), [link](https://github.com/dotCMS/core/pull/25138/files?diff=unified&w=0#diff-31b7a396b598816d0a94599edd0c1338d80f276fed9775642769710040776e24L221-R219), [link](https://github.com/dotCMS/core/pull/25138/files?diff=unified&w=0#diff-a32a0fcbcf1b026389eade1b0cd88e8838bbfacb92905fe9da2bea7f72b7c9abL107-R107), [link](https://github.com/dotCMS/core/pull/25138/files?diff=unified&w=0#diff-a32a0fcbcf1b026389eade1b0cd88e8838bbfacb92905fe9da2bea7f72b7c9abL169-R191), [link](https://github.com/dotCMS/core/pull/25138/files?diff=unified&w=0#diff-d12c3fc3a47505ab4b846cfc5090445b7597e2acd4c10382ff0dc22f8c0d9ff0L80-R81), [link](https://github.com/dotCMS/core/pull/25138/files?diff=unified&w=0#diff-d12c3fc3a47505ab4b846cfc5090445b7597e2acd4c10382ff0dc22f8c0d9ff0L130-R131))
*  Update `createDotGridStackWidgetFromNode` function to accept a second parameter `isNewNode` and generate a new id for the widget if it is a new node ([link](https://github.com/dotCMS/core/pull/25138/files?diff=unified&w=0#diff-a32a0fcbcf1b026389eade1b0cd88e8838bbfacb92905fe9da2bea7f72b7c9abL107-R107), [link](https://github.com/dotCMS/core/pull/25138/files?diff=unified&w=0#diff-f3792a46b295fb618b6d4a8d8f4ecef72ec5e3c41a470d6a44886881a6b93659L59-R65))
*  Add `removeColumn` method to `TemplateBuilderComponent` to handle the logic of removing a column from the grid and the store ([link](https://github.com/dotCMS/core/pull/25138/files?diff=unified&w=0#diff-d12c3fc3a47505ab4b846cfc5090445b7597e2acd4c10382ff0dc22f8c0d9ff0L6-R7), [link](https://github.com/dotCMS/core/pull/25138/files?diff=unified&w=0#diff-d12c3fc3a47505ab4b846cfc5090445b7597e2acd4c10382ff0dc22f8c0d9ff0R150-R171))
*  Add `width: 100%` style to the `template-builder-box` class and remove `width: $spacing-9` style from the `template-builder-box--small` class to make the box component responsive to the grid size ([link](https://github.com/dotCMS/core/pull/25138/files?diff=unified&w=0#diff-730c873d53a5e9ee2f15dba5eb5535811fe305d95b0ea744b18f44a055ef14f5R23), [link](https://github.com/dotCMS/core/pull/25138/files?diff=unified&w=0#diff-730c873d53a5e9ee2f15dba5eb5535811fe305d95b0ea744b18f44a055ef14f5L59-R61))
*  Remove `padding: $spacing-3 0` style from the `template-builder-row` class to avoid extra space between the rows in the grid ([link](https://github.com/dotCMS/core/pull/25138/files?diff=unified&w=0#diff-849ff937618b9b9a9847737f9332ad3f6d2ddd0e3fb6f037cdef6abd32483a4dL34))
*  Change `cellHeight` property of the `subGridOptions` object from `224` to `257` to add some padding to the subgrid cells ([link](https://github.com/dotCMS/core/pull/25138/files?diff=unified&w=0#diff-e6ced9e93037bc435963ea7ee26af5807e76d953f74512e9e546bf1e5468fa3eL33-R33))
*  Add `data-testId` attributes to the add row button, the add box button, and the row element in the template of `TemplateBuilderComponent` to facilitate the testing of the component ([link](https://github.com/dotCMS/core/pull/25138/files?diff=unified&w=0#diff-1fde9398eccbbd88fe2b6ce2c5d40bcefeb93c5bc3c7f110e3a1ed6b29b32aa3R8), [link](https://github.com/dotCMS/core/pull/25138/files?diff=unified&w=0#diff-1fde9398eccbbd88fe2b6ce2c5d40bcefeb93c5bc3c7f110e3a1ed6b29b32aa3R16), [link](https://github.com/dotCMS/core/pull/25138/files?diff=unified&w=0#diff-1fde9398eccbbd88fe2b6ce2c5d40bcefeb93c5bc3c7f110e3a1ed6b29b32aa3R29))
*  Remove the entire content of the `template-builder.component.scss` file, since the styles were moved to the respective components or were no longer needed ([link](https://github.com/dotCMS/core/pull/25138/files?diff=unified&w=0#diff-5e1e17de0055cdeac5f6182ffbe2a5a22026c3d8b74bb7ecc7dea4d217cf69ceL3-L16))
*  Replace the entire content of the `template-builder.component.spec.ts` file with a new test suite that uses the host factory to create a host component with the `TemplateBuilderComponent` as a child component ([link](https://github.com/dotCMS/core/pull/25138/files?diff=unified&w=0#diff-71156e8bc7aa2247cecab3f2627182871f930b385d4ff4de19df2f0951667597L1-R6), [link](https://github.com/dotCMS/core/pull/25138/files?diff=unified&w=0#diff-71156e8bc7aa2247cecab3f2627182871f930b385d4ff4de19df2f0951667597L10-R65))
*  Add `TemplateBuilderBoxComponent` to the imports array of the module metadata decorator of the `TemplateBuilderComponent` story in the `template-builder.component.stories.ts` file ([link](https://github.com/dotCMS/core/pull/25138/files?diff=unified&w=0#diff-cbdb877fac9228e73e54cf3af3465499a71c93f2c594f45769081d65d2881653R6), [link](https://github.com/dotCMS/core/pull/25138/files?diff=unified&w=0#diff-cbdb877fac9228e73e54cf3af3465499a71c93f2c594f45769081d65d2881653L16-R23))

### Screenshots
https://github.com/dotCMS/core/assets/63567962/542d7ca7-2e46-4749-9e4d-efcf0944bb4f

### CODE REVIEW

File: core-web/libs/template-builder/src/lib/components/template-builder/utils/gridstack-options.ts [hunk 1]
```diff
@@ -30,7 +30,7 @@ function isARowWidget(el: Element): boolean {
 }
 
 export const subGridOptions: GridStackOptions = {
-    cellHeight: 224,
+    cellHeight: 257, // 32px more to make the padding
```
Comment (fmontes) [inline core-web/libs/template-builder/src/lib/components/template-builder/utils/gridstack-options.ts:33]:
```
const BOX_PADDNG = 32;
```

```suggestion
    cellHeight: 224 + BOX_PADDNG
```

File: core-web/libs/template-builder/src/lib/components/template-builder/utils/gridstack-utils.ts [hunk 2]
```diff
@@ -56,10 +56,13 @@ export function createDotGridStackWidgets(columns: DotGridStackNode[]): DotGridS
...
  */
-export function createDotGridStackWidgetFromNode(node: DotGridStackNode): DotGridStackWidget {
+export function createDotGridStackWidgetFromNode(
+    node: DotGridStackNode,
+    isNewNode: boolean = false
+): DotGridStackWidget {
     return {
         x: node.x,
-        id: uuid(),
+        id: isNewNode ? uuid() : node.id,
```
Comment (fmontes) [inline core-web/libs/template-builder/src/lib/components/template-builder/utils/gridstack-utils.ts:65]:
```suggestion
        id: node.id || uuid(),
```

**Review by fmontes — CHANGES_REQUESTED** [review]

**Review by rjvelazco — APPROVED** [review]

File: core-web/libs/template-builder/src/lib/components/template-builder/components/template-builder-box/template-builder-box.component.scss [hunk 3]
```diff
@@ -20,6 +20,8 @@
     border: 1px solid $color-palette-primary-300;
     border-radius: $border-radius-sm;
     height: 14rem;
+    width: 100%;
+    background-color: white;
```
Comment (fmontes) [inline core-web/libs/template-builder/src/lib/components/template-builder/components/template-builder-box/template-builder-box.component.scss:24]:
```suggestion
    background-color: $white;
```

File: core-web/libs/template-builder/src/lib/components/template-builder/template-builder.component.scss [hunk 4]
```diff
@@ -29,3 +15,10 @@ dotcms-add-widget.ui-draggable-dragging {
     transform: rotate(-3.66deg);
     box-shadow: $md-shadow-6;
 }
+
+.ui-draggable-dragging {
+    dotcms-template-builder-row {
+        background-color: white;
```
Comment (fmontes) [inline core-web/libs/template-builder/src/lib/components/template-builder/template-builder.component.scss:21]:
```suggestion
        background-color: $white;
```

**Review by fmontes — CHANGES_REQUESTED** [review]

**Review by manuelrojas — APPROVED** [review]

File: core-web/libs/template-builder/src/lib/components/template-builder/components/template-builder-box/template-builder-box.component.ts [hunk 5]
```diff
@@ -1,28 +1,45 @@
...
     selector: 'dotcms-template-builder-box',
     templateUrl: './template-builder-box.component.html',
     styleUrls: ['./template-builder-box.component.scss'],
     changeDetection: ChangeDetectionStrategy.OnPush,
     standalone: true,
-    imports: [NgFor, NgForOf, NgIf, NgClass, ButtonModule, ScrollPanelModule, DotMessagePipeModule]
+    imports: [
+        NgFor,
+        NgForOf,
+        NgIf,
```
Comment (fmontes) [inline core-web/libs/template-builder/src/lib/components/template-builder/components/template-builder-box/template-builder-box.component.ts:32]: You sure we need all these?

File: core-web/libs/template-builder/src/lib/components/template-builder/components/template-builder-box/template-builder-box.component.ts [hunk 6]
```diff
@@ -31,9 +48,22 @@ export class TemplateBuilderBoxComponent {
...
+    @Output()
+    deleteColumnRejected: EventEmitter<void> = new EventEmitter<void>();
 
     @Input() items: DotTemplateBuilderContainer[];
 
-    protected readonly templateBuilderSizes = TemplateBuilderBoxSize;
-    @Input() size: TemplateBuilderBoxSize = TemplateBuilderBoxSize.large;
+    @Input() size = 1;
+
+    boxSize = TemplateBuilderBoxSize.small;
```
Comment (fmontes) [inline core-web/libs/template-builder/src/lib/components/template-builder/components/template-builder-box/template-builder-box.component.ts:58]: Why size and boxSize? seems redundant to me.

File: core-web/libs/template-builder/src/lib/components/template-builder/template-builder.component.spec.ts [hunk 7]
```diff
@@ -27,25 +41,70 @@ describe('TemplateBuilderComponent', () => {
...
                 }
             }
+        );
+
+        store = spectator.inject(DotTemplateBuilderStore);
+    });
+
+    it('should create', () => {
+        expect(spectator).toBeTruthy();
+    });
```
Comment (fmontes) [inline core-web/libs/template-builder/src/lib/components/template-builder/template-builder.component.spec.ts:59]: Not needed.

File: core-web/libs/template-builder/src/lib/components/template-builder/template-builder.component.spec.ts [hunk 8]
```diff
@@ -27,25 +41,70 @@ describe('TemplateBuilderComponent', () => {
...
+    it('should have a Add Row Button', () => {
+        expect(spectator.query(byTestId('add-row'))).toBeTruthy();
+    });
+
+    it('should have a Add Box Button', () => {
+        expect(spectator.query(byTestId('add-box'))).toBeTruthy();
+    });
+
+    it('should have a 12 columns as background', () => {
+        expect(spectator.queryAll(byTestId('column')).length).toBe(12);
```
Comment (fmontes) [inline core-web/libs/template-builder/src/lib/components/template-builder/template-builder.component.spec.ts:70]: This should be inside the component that draw this, here we just need to look for that component.

File: core-web/libs/template-builder/src/lib/components/template-builder/utils/gridstack-options.ts [hunk 9]
```diff
@@ -36,21 +51,22 @@ function isARowWidget(el: Element): boolean {
 }
 
 export const subGridOptions: GridStackOptions = {
-    cellHeight: 224,
+    cellHeight: GRID_STACK_BOX_HEIGHT,
     column: 'auto',
-    margin: `${GRID_STACK_MARGIN}${GRID_STACK_UNIT}`,
+    margin: `${GRID_STACK_MARGIN_VERTICAL}${GRID_STACK_UNIT} ${GRID_STACK_MARGIN_HORIZONTAL}${GRID_STACK_UNIT}`,
```
Comment (fmontes) [inline core-web/libs/template-builder/src/lib/components/template-builder/utils/gridstack-options.ts:56]: This takes a `marginUnit` I think we should use it.

**Review by fmontes — CHANGES_REQUESTED** [review]

File: core-web/libs/template-builder/src/lib/components/template-builder/components/template-builder-box/template-builder-box.component.ts (hunk 6 above)
Comment (zJaaal) [reply core-web/libs/template-builder/src/lib/components/template-builder/components/template-builder-box/template-builder-box.component.ts:58]: agree, changing to `boxVariant`

File: core-web/libs/template-builder/src/lib/components/template-builder/utils/gridstack-options.ts (hunk 9 above)
Comment (zJaaal) [reply core-web/libs/template-builder/src/lib/components/template-builder/utils/gridstack-options.ts:56]: That's right, but it says that `marginUnit` property is an internal, and it setted by GridStack when the margin is a `string`

**Review by fmontes — APPROVED** [review]

**Pull request merged**